Ocpp: segfault

### Describe the bug

[main  ] INFO 2023/06/24 17:13:40 evcc 0.118.1
[main  ] INFO 2023/06/24 17:13:40 using config file: /etc/evcc.yaml
[main  ] INFO 2023/06/24 17:13:40 starting ui and api at :7070
[db    ] INFO 2023/06/24 17:13:41 using sqlite database: /home/pi/.evcc/evcc.db
[ocpp  ] WARN 2023/06/24 17:13:44 chargepoint connected, unknown: TACWxxxxxx
[ocpp  ] DEBUG 2023/06/24 17:13:44 waiting for chargepoint: 5m0s
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1341fa0]

goroutine 30 [running]:
github.com/evcc-io/evcc/charger/ocpp.(*CP).StatusNotification(0x0, 0x4ebd110)
	/home/runner/work/evcc/evcc/charger/ocpp/cp_core.go:54 +0x4c
github.com/evcc-io/evcc/charger/ocpp.(*CS).OnStatusNotification(0x54ff4b8, {0x5835465, 0x10}, 0x4ebd110)
	/home/runner/work/evcc/evcc/charger/ocpp/cs_core.go:106 +0x68
github.com/lorenzodonini/ocpp-go/ocpp1%2e6.(*centralSystem).handleIncomingRequest.func1()
	/home/runner/go/pkg/mod/github.com/lorenzodonini/ocpp-go@v0.16.1-0.20230416102136-782920b66533/ocpp1.6/central_system.go:519 +0x598
created by github.com/lorenzodonini/ocpp-go/ocpp1%2e6.(*centralSystem).handleIncomingRequest
	/home/runner/go/pkg/mod/github.com/lorenzodonini/ocpp-go@v0.16.1-0.20230416102136-782920b66533/ocpp1.6/central_system.go:502 +0x508


### Steps to reproduce

start evcc

### Configuration details

```yaml
-
```


### Log details

```text
-
```


### What type of operating system are you running?

Linux

### Version

0.118.1

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -75,9 +75,7 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 
 		// register unknown chargepoint
 		// when chargepoint setup is complete, it will eventually be associated with the connected id
-		cs.mu.Lock()
 		cs.cps[chargePoint.ID()] = nil
-		cs.mu.Unlock()
 	} else {
 		cs.log.DEBUG.Printf("chargepoint connected: %s", chargePoint.ID())
 

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -44,6 +44,9 @@ func (cs *CS) errorHandler(errC <-chan error) {
 }
 
 func (cs *CS) chargepointByID(id string) (*CP, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, ok := cs.cps[id]
 	if !ok {
 		return nil, fmt.Errorf("unknown charge point: %s", id)
@@ -52,10 +55,10 @@ func (cs *CS) chargepointByID(id string) (*CP, error) {
 }
 
 func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
-	cs.mu.Lock()
-	defer cs.mu.Unlock()
-
 	if cp, err := cs.chargepointByID(chargePoint.ID()); err != nil {
+		cs.mu.Lock()
+		defer cs.mu.Unlock()
+
 		// check for anonymous chargepoint
 		if cp, ok := cs.cps[""]; ok {
 			cs.log.INFO.Printf("chargepoint connected, registering: %s", chargePoint.ID())
@@ -86,9 +89,6 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 }
 
 func (cs *CS) ChargePointDisconnected(chargePoint ocpp16.ChargePointConnection) {
-	cs.mu.Lock()
-	defer cs.mu.Unlock()
-
 	if cp, err := cs.chargepointByID(chargePoint.ID()); err != nil {
 		cs.log.ERROR.Printf("chargepoint disconnected: %v", err)
 	} else {
@@ -96,7 +96,9 @@ func (cs *CS) ChargePointDisconnected(chargePoint ocpp16.ChargePointConnection) 
 
 		if cp == nil {
 			// remove unknown chargepoint
+			cs.mu.Lock()
 			delete(cs.cps, chargePoint.ID())
+			cs.mu.Unlock()
 		} else {
 			cp.connect(false)
 		}

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -44,9 +44,6 @@ func (cs *CS) errorHandler(errC <-chan error) {
 }
 
 func (cs *CS) chargepointByID(id string) (*CP, error) {
-	cs.mu.Lock()
-	defer cs.mu.Unlock()
-
 	cp, ok := cs.cps[id]
 	if !ok {
 		return nil, fmt.Errorf("unknown charge point: %s", id)
@@ -55,6 +52,9 @@ func (cs *CS) chargepointByID(id string) (*CP, error) {
 }
 
 func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	if cp, err := cs.chargepointByID(chargePoint.ID()); err != nil {
 		// check for anonymous chargepoint
 		if cp, err := cs.chargepointByID(""); err == nil {
@@ -63,10 +63,8 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 			// update id
 			cp.RegisterID(chargePoint.ID())
 
-			cs.mu.Lock()
 			cs.cps[chargePoint.ID()] = cp
 			delete(cs.cps, "")
-			cs.mu.Unlock()
 
 			cp.connect(true)
 
@@ -91,6 +89,9 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 }
 
 func (cs *CS) ChargePointDisconnected(chargePoint ocpp16.ChargePointConnection) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	if cp, err := cs.chargepointByID(chargePoint.ID()); err != nil {
 		cs.log.ERROR.Printf("chargepoint disconnected: %v", err)
 	} else {
@@ -98,9 +99,7 @@ func (cs *CS) ChargePointDisconnected(chargePoint ocpp16.ChargePointConnection) 
 
 		if cp == nil {
 			// remove unknown chargepoint
-			cs.mu.Lock()
 			delete(cs.cps, chargePoint.ID())
-			cs.mu.Unlock()
 		} else {
 			cp.connect(false)
 		}

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -56,17 +56,17 @@ func (cs *CS) chargepointByID(id string) (*CP, error) {
 
 func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 	if cp, err := cs.chargepointByID(chargePoint.ID()); err != nil {
-		cs.mu.Lock()
-		defer cs.mu.Unlock()
-
 		// check for anonymous chargepoint
-		if cp, ok := cs.cps[""]; ok {
+		if cp, err := cs.chargepointByID(""); err == nil {
 			cs.log.INFO.Printf("chargepoint connected, registering: %s", chargePoint.ID())
 
 			// update id
 			cp.RegisterID(chargePoint.ID())
+
+			cs.mu.Lock()
 			cs.cps[chargePoint.ID()] = cp
 			delete(cs.cps, "")
+			cs.mu.Unlock()
 
 			cp.connect(true)
 
@@ -77,7 +77,9 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 
 		// register unknown chargepoint
 		// when chargepoint setup is complete, it will eventually be associated with the connected id
+		cs.mu.Lock()
 		cs.cps[chargePoint.ID()] = nil
+		cs.mu.Unlock()
 	} else {
 		cs.log.DEBUG.Printf("chargepoint connected: %s", chargePoint.ID())
 

--- a/charger/ocpp/cs_core.go
+++ b/charger/ocpp/cs_core.go
@@ -53,6 +53,9 @@ func (cs *CS) TriggerMeterValuesRequest(id string, connector int) {
 // cp actions
 
 func (cs *CS) OnAuthorize(id string, request *core.AuthorizeRequest) (*core.AuthorizeConfirmation, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -62,6 +65,9 @@ func (cs *CS) OnAuthorize(id string, request *core.AuthorizeRequest) (*core.Auth
 }
 
 func (cs *CS) OnBootNotification(id string, request *core.BootNotificationRequest) (*core.BootNotificationConfirmation, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -71,6 +77,9 @@ func (cs *CS) OnBootNotification(id string, request *core.BootNotificationReques
 }
 
 func (cs *CS) OnDataTransfer(id string, request *core.DataTransferRequest) (*core.DataTransferConfirmation, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -80,6 +89,9 @@ func (cs *CS) OnDataTransfer(id string, request *core.DataTransferRequest) (*cor
 }
 
 func (cs *CS) OnHeartbeat(id string, request *core.HeartbeatRequest) (*core.HeartbeatConfirmation, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -89,6 +101,9 @@ func (cs *CS) OnHeartbeat(id string, request *core.HeartbeatRequest) (*core.Hear
 }
 
 func (cs *CS) OnMeterValues(id string, request *core.MeterValuesRequest) (*core.MeterValuesConfirmation, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -98,6 +113,9 @@ func (cs *CS) OnMeterValues(id string, request *core.MeterValuesRequest) (*core.
 }
 
 func (cs *CS) OnStatusNotification(id string, request *core.StatusNotificationRequest) (*core.StatusNotificationConfirmation, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -107,6 +125,9 @@ func (cs *CS) OnStatusNotification(id string, request *core.StatusNotificationRe
 }
 
 func (cs *CS) OnStartTransaction(id string, request *core.StartTransactionRequest) (*core.StartTransactionConfirmation, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -116,6 +137,9 @@ func (cs *CS) OnStartTransaction(id string, request *core.StartTransactionReques
 }
 
 func (cs *CS) OnStopTransaction(id string, request *core.StopTransactionRequest) (*core.StopTransactionConfirmation, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -125,6 +149,9 @@ func (cs *CS) OnStopTransaction(id string, request *core.StopTransactionRequest)
 }
 
 func (cs *CS) OnDiagnosticsStatusNotification(id string, request *firmware.DiagnosticsStatusNotificationRequest) (confirmation *firmware.DiagnosticsStatusNotificationConfirmation, err error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err
@@ -134,6 +161,9 @@ func (cs *CS) OnDiagnosticsStatusNotification(id string, request *firmware.Diagn
 }
 
 func (cs *CS) OnFirmwareStatusNotification(id string, request *firmware.FirmwareStatusNotificationRequest) (confirmation *firmware.FirmwareStatusNotificationConfirmation, err error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
 	cp, err := cs.chargepointByID(id)
 	if err != nil {
 		return nil, err

